compiler/rustc_index/src/vec.rs

```diff
@@ -12,7 +12,19 @@ use std::vec;
 use crate::{Idx, IndexSlice};
 
 /// An owned contiguous collection of `T`s, indexed by `I` rather than by `usize`.
-/// Its purpose is to avoid mixing indexes.
+///
+/// Why use this instead of a `Vec`?
+/// This enforces the user to index a given IndexVec only with the associated index thus making it
```

Comment:
While it does *enforce* that it's used correctly, I think the way more valuable part is *documenting* what the index is. When I see `HashMap<usize, Something>` it's not immediately clear what the `usize` means, while `HashMap<FieldIdx, Something>` makes it obvious.